### DISCUSSION TITLE
An appeal for test driven development!

### DISCUSSION THREAD

**Discussion by jbusecke** [GliderToolsCommunity/GliderTools#114] [Show and tell]
Hi everyone, 

Since we are seeing some PRs coming in (thanks @callumrollo @isgiddy ) I just wanted to layout a general strategy that hopefully will help all of us to increase the test coverage (and make development easier actually).

I would strongly encourage '*test-driven development*', which basically means that for any change you make, you should think about and write the test **before** the actual code. 

For a bug fix this usually looks like this: 
1. You have found an error. The first step is to modify/add a test for the failing function that reproduces the error.  This should fail on purpose by expecting the correct output! This way you are sure that this exact bug will be checked for in every future change to the code.
2. Now you start modifying the code until the test passes.

Same for a new feature: 
1. Write the test first, thinking about all scenarios and what output you would expect for a given input data.
2. Then implement the actual feature in the code.

This might seem a bit counterintuitive at first, but I am very convinced that this encourages a lot better code, and makes the maintenance easier.

Let me know if anything about this is unclear. And keep those amazing PRs coming! Hehe

**Comment by Thomas-Moore-Creative** [comment]
@jbusecke : thanks for your efforts!  Do you have any advice on where a TDD noob could go to get more basic advice and examples on  best-practices for approaching TDD?

**Reply by jbusecke** [reply]
I am in the process of writing a paper/guide on exactly that, and will share that as soon as possible!
